grafana/resource_team_external_group.go

```diff
@@ -0,0 +1,165 @@
+* [HTTP API](https://grafana.com/docs/grafana/latest/http_api/external_group_sync/)
+`,
+
+		CreateContext: CreateTeamExternalGroup,
+		UpdateContext: UpdateTeamExternalGroup,
+		DeleteContext: DeleteTeamExternalGroup,
+		ReadContext:   ReadTeamExternalGroup,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"team_id": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The Team ID",
+			},
+
+			"groups": {
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Description: `The team external groups list`,
+			},
+		},
+	}
+}
+
+func CreateTeamExternalGroup(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	teamID := d.Get("team_id").(int)
+	d.SetId(strconv.FormatInt(int64(teamID), 10))
+	if err := UpdateGroups(d, meta); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diag.Diagnostics{}
+}
+
+func ReadTeamExternalGroup(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	if err := ReadGroups(d, meta); err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func UpdateTeamExternalGroup(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	if err := UpdateGroups(d, meta); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diag.Diagnostics{}
+}
+
+func DeleteTeamExternalGroup(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	if err := UpdateGroups(d, meta); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diag.Diagnostics{}
+}
+
+func ReadGroups(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*client).gapi
+	teamID, _ := strconv.ParseInt(d.Id(), 10, 64)
+	teamGroups, err := client.TeamGroups(teamID)
+	if err != nil {
+		return err
+	}
+	groupSlice := []string{}
+	for _, teamGroup := range teamGroups {
+		groupSlice = append(groupSlice, teamGroup.GroupID)
+	}
+	d.Set("groups", groupSlice)
+
+	return nil
+}
+
+func UpdateGroups(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*client).gapi
+
+	addGroups, removeGroups := groupChanges(d)
+	teamID, _ := strconv.ParseInt(d.Id(), 10, 64)
+
+	for _, group := range addGroups {
+		err := client.NewTeamGroup(teamID, group)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, group := range removeGroups {
+		err := client.DeleteTeamGroup(teamID, group)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func groupChanges(d *schema.ResourceData) ([]string, []string) {
+	// Get the lists of team groups read in from Grafana state (old) and configured (new)
+	state, config := d.GetChange("groups")
+
+	stateGroups := make([]string, state.(*schema.Set).Len())
+	for _, v := range state.(*schema.Set).List() {
+		stateGroups = append(stateGroups, v.(string))
+	}
```

Comment:
Here and below when you declare the slice it will already have `len` empty elements, so when you append to them you will end up with twice the size of the slice, as shown in [here](https://play.golang.org/p/2tuFhXP7Vni).

There are two ways you can fix this: if you want to use append, then initialise the slice as `stateGroups := make([]string, 0, state.(*schema.Set).Len())`, so it has the right capacity but no elements assigned. Or range with an index and access the destination elements in the slice directly:

```suggestion
	stateGroups := make([]string, state.(*schema.Set).Len())
	for i, v := range state.(*schema.Set).List() {
		stateGroups[i] = v.(string)
	}
```